src/core/server/rendering/create_metadata_factory.ts

```diff
@@ -0,0 +1,128 @@
+  params: RenderingProviderParams
+) {
+  const config = (params.http.server as any).config();
+  const defaultVars =
+    params.legacyPlugins.uiExports.defaultInjectedVarProviders?.reduce(
+      (defaults, { fn, pluginSpec }) =>
+        mergeVariables(fn(params.http.server, pluginSpec.readConfigValue(config, [])), defaults),
+      {}
+    ) ?? {};
+  const injectedVars = mergeVariables(
+    defaultVars,
+    ...pluginConfigs,
+    await params.getVarsFor(id),
+    params.injectedVarsOverrides || {}
+  );
+
+  return (
+    params.legacyPlugins.uiExports.injectedVarsReplacers?.reduce(
+      async (vars, replacer) => await replacer(vars, params.request, params.http.server),
+      injectedVars
+    ) ?? injectedVars
+  );
+}
+
+function getUiPlugins(params: RenderingProviderParams) {
+  return Promise.all(
+    [...params.plugins.uiPlugins.public].map(async ([pluginId, plugin]) => {
+      const config = (await params.plugins.uiPlugins.browserConfigs
+        .get(pluginId)
+        ?.pipe(take(1))
+        .toPromise()) as PluginVariables;
+
+      return { id: pluginId, plugin, config: config || {} };
+    })
+  );
+}
+
+async function getSettings(params: RenderingProviderParams, includeUserProvidedConfig: boolean) {
+  return {
+    defaults: await params.uiSettings.getAll(),
+    user: includeUserProvidedConfig ? await params.uiSettings.getUserProvided() : {},
+  };
+}
+
+/**
+ * Returns a function that generates all the metadata necessary for rendering a boostrapped page.
+ */
+export const createMetadataFactory = (params: RenderingProviderParams) =>
+  /**
+   * Generate all the metadata necessary for rendering a boostrapped page.
+   */
+  async (id = 'core', includeUserProvidedConfig = true): Promise<RenderingMetadata> => {
+    const basePath = params.http.basePath.get(params.request);
+    const uiPlugins = await getUiPlugins(params);
+    const settings = await getSettings(params, includeUserProvidedConfig);
+    const vars = await getVariables(
+      id,
+      uiPlugins.map(({ config }) => config),
+      params
+    );
+
+    return {
+      strictCsp: params.http.csp.strict,
+      uiPublicUrl: `${basePath}/ui`,
+      bootstrapScriptUrl: `${basePath}/bundles/app/${id}/bootstrap.js`,
+      i18n: i18n.translate,
+      locale: i18n.getLocale(),
+      darkMode: settings.user?.['theme:darkMode']?.userValue
+        ? Boolean(settings.user['theme:darkMode'].userValue)
+        : false,
+      injectedMetadata: {
+        version: params.env.packageInfo.version,
+        buildNumber: params.env.packageInfo.buildNum,
+        branch: params.env.packageInfo.branch,
+        basePath,
+        env: params.env,
+        legacyMode: id !== 'core',
+        i18n: {
+          translationsUrl: `${basePath}/translations/${i18n.getLocale()}.json`,
+        },
+        csp: { warnLegacyBrowsers: params.http.csp.warnLegacyBrowsers },
+        vars,
+        uiPlugins,
+        legacyMetadata: {
+          app: {},
```

Comment:
is `app` required or not?

Reply:
I couldn't find any usages, but I wasn't sure if this was used elsewhere. I'll remove an add a note that it has been removed.